lib/core/src/Cardano/DB/Sqlite.hs

```diff
@@ -167,9 +179,82 @@ queryLogFunc tr _loc _source level str = traceWith tr (MsgQuery msg sev)
-      select _ = Nothing
-      handler = const . pure  . Left $ e
+    select (SqliteException ErrorConstraint _ _) = Just ()
+    select _ = Nothing
+    handler = const . pure  . Left $ e
+
+{-------------------------------------------------------------------------------
+                           Internal / Database Setup
+-------------------------------------------------------------------------------}
+
+newInMemorySqliteContext
+    :: Tracer IO DBLog
+    -> [ManualMigration]
+    -> Migration
+    -> IO SqliteContext
+newInMemorySqliteContext tr manualMigrations autoMigration = do
+    conn <- Sqlite.open connStr
+    mapM_ (`executeManualMigration` conn) manualMigrations
+    unsafeBackend <- wrapConnectionInfo info conn (queryLogFunc tr)
+    void $ runSqlConn (runMigrationQuiet autoMigration) unsafeBackend
+
+    let observe :: forall a. IO a -> IO a
+        observe = bracketTracer (contramap MsgRun tr)
+
+    -- We still use a lock with the in-memory database to protect it from
+    -- concurrent accesses and ensure database integrity in case where multiple
+    -- threads would be reading/writing from/to it.
+    lock <- newMVar unsafeBackend
+    let runQuery :: forall a. SqlPersistT IO a -> IO a
+        runQuery cmd = withMVarMasked lock (observe . runSqlConn cmd)
+
+    return $ SqliteContext { runQuery, dbFile }
+  where
+    dbFile = Nothing
+    connStr = sqliteConnStr dbFile
+    info = mkSqliteConnectionInfo connStr
+
+-- | Sets up query logging and timing, runs schema migrations if necessary and
+-- provide a safe 'SqliteContext' for interacting with the database.
+newSqliteContext
+    :: Tracer IO DBLog
+    -> ConnectionPool
+    -> [ManualMigration]
+    -> Migration
+    -> FilePath
+    -> IO (Either MigrationError SqliteContext)
+newSqliteContext tr pool manualMigrations autoMigration fp = do
+    migrationResult <- withResource pool $ \(backend, conn) -> do
+        let executeAutoMigration = runSqlConn (runMigrationQuiet autoMigration) backend
+        migrationResult <- withForeignKeysDisabled tr conn $ do
+            mapM_ (`executeManualMigration` conn) manualMigrations
+            executeAutoMigration
+                & tryJust (matchMigrationError @PersistException)
+                & tryJust (matchMigrationError @SqliteException)
+                & fmap join
+        traceWith tr $ MsgMigrations $ fmap length migrationResult
+        return migrationResult
+    return $ case migrationResult of
+        Left e  -> Left e
+        Right{} ->
+            let observe :: IO a -> IO a
+                observe = bracketTracer (contramap MsgRun tr)
+
+               -- runSqlConn is guarded with a lock because it's not threadsafe in
+               -- general.It is also masked, so that the SqlBackend state is not
+               -- corrupted if a thread gets cancelled while running a query.
+               -- See: https://github.com/yesodweb/persistent/issues/981
```

Comment:
Perhaps in future we could drop the lock and mask_?
1. If we are confident that any pool connection will only be used by one thread at a time, then the lock isn't needed.
2. If the thread is cancelled during a database operation, then so be it. SQLite will rollback the uncommitted transaction. The connection won't be returned to the pool.

Reply:
(1) is only true if we can start the database in full_mutex I guess, which persistent doesn't really allow us to set that flag (although, we've forked it so it's easily added). This will not prevent retries however in case two writing threads are trying to write something at the same time, but at least it should give more concurrency to readers :thinking: Also "if we are confident that any pool connection will only be used by one thread at a time" ==> I have no such confidence at this stage unfortunately :( 

(2) Hmmm. That's right. Sounds scary though :scream: